gpflow/experimental/check_shapes/__init__.py

```diff
@@ -48,247 +42,282 @@ def linear_model(
+   :dedent:
 
 
 Shape reuse
 +++++++++++
 
 Just like with other code it is useful to be able to specify a shape in one place and reuse the
-specification. In particular this ensures that your code keep having consistent shapes, even if it
-is refactored.
+specification. In particular this ensures that your code keep having internally consistent shapes,
+even if it is refactored.
 
 
 Class inheritance
 -----------------
 
 If you have a class hiererchy, you probably want to ensure that derived classes handle tensors with
 the same shapes as the base classes. You can use the :func:`inherit_check_shapes` decorator to
-inherit shapes from overridden methods.
-
-Example::
+inherit shapes from overridden methods:
 
-    class SuperClass(ABC):
-        @abstractmethod
-        @check_shapes(
-            "a: [batch..., 4]",
-            "return: [batch..., 1]",
-        )
-        def f(self, a: tf.Tensor) -> tf.Tensor:
-            ...
-
-    class SubClass(SuperClass):
-        @inherit_check_shapes
-        def f(self, a: tf.Tensor) -> tf.Tensor:
-            ...
+.. literalinclude:: /examples/test_check_shapes_examples.py
+   :start-after: [reuse__inherit_check_shapes]
+   :end-before: [reuse__inherit_check_shapes]
+   :dedent:
 
 
 Functional programming
 ----------------------
 
 If you prefer functional- over object oriented programming, you may have functions that you require
 to handle the same shapes. To do this, remember that in Python a decorator is just a function, and
-functions are objects that can be stored::
-
-    check_my_shapes = check_shapes(
-        "a: [batch..., 4]",
-        "return: [batch..., 1]",
-    )
+functions are objects that can be stored:
 
-    @check_my_shapes
-    def f(a: tf.Tensor) -> tf.Tensor:
-        ...
-
-    @check_my_shapes
-    def g(a: tf.Tensor) -> tf.Tensor:
-        ...
+.. literalinclude:: /examples/test_check_shapes_examples.py
+   :start-after: [reuse__functional]
+   :end-before: [reuse__functional]
+   :dedent:
 
 
 Other reuse of shapes
 ---------------------
 
 You can use :func:`get_check_shapes` to get, and reuse, the shape definitions from a previously
 declared function. This is particularly useful to ensure fakes in tests use the same shapes as the
-production implementation::
-
-    @check_shapes(
-        "a: [batch..., 4]",
-        "return: [batch..., 1]",
-    )
-    def f(a: tf.Tensor) -> tf.Tensor:
-        ...
+production implementation:
 
-    def test_something() -> None:
-        @get_check_shapes(f)
-        def fake_f(a: tf.Tensor) -> tf.Tensor:
-            ...
-
-        # Test that patches `f` with `fake_f` goes here...
+.. literalinclude:: /examples/test_check_shapes_examples.py
+   :start-after: [reuse__get_check_shapes]
+   :end-before: [reuse__get_check_shapes]
+   :dedent:
 
 
 Checking shapes without a decorator
 +++++++++++++++++++++++++++++++++++
 
 While the :func:`check_shapes` decorator is the recommend way to use this library, it is possible to
```

Comment:
why are we exposing this in the public API? Maybe best to keep hidden from the user in case you want to make changes to this later?

Reply:
> why are we exposing this in the public API? Maybe best to keep hidden from the user in case you want to make changes to this later?

Hmmm, maybe. Not sure. I like giving people a fallback in case there is some reason they cannot use the decorator. As you've seen yourself, using decorators can sometimes give weird side-effects, so I don't like forcing people to use them.